utils/ptrace.c

```diff
@@ -0,0 +1,129 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ptrace.h>
+#include <sys/user.h>
+#include <wait.h>
+#include <time.h>
+
+#include "ptrace.h"
+#include "utils.h"
+
+void ptrace_attach(pid_t target)
+{
+	int waitpidstatus;
+
+	if (ptrace(PTRACE_ATTACH, target, NULL, NULL) == -1)
+		pr_err("ptrace(PTRACE_ATTACH) failed");
+
+	if (waitpid(target, &waitpidstatus, WUNTRACED) != target)
```

Comment:
I'm not sure `WUNTRACED` flag is needed.  Also it seems you need to check the status (`WSTOPSIG(status) == SIGSTOP`).

Reply:
And I think you'd be better stopping other threads first.  As I read the ptrace manual, after `PTRACE_ATTACH` you'll need to inject SIGSTOP with `PTRACE_CONT`.

Reply:
@namhyung PTRACE_ATTACH sends SIGSTOP to target. target process and its childs will stop when it have recieved SIGSTOP. so, I think this code reasonable.

Reply:
Yes `PTRACE_ATTACH` will send it but after signal-delivery-stop you didn't inject it to the target process.  Please read "Attaching and detaching" section in the ptrace man page.